pkg/integrations/v2/eventhandler/eventhandler.go

```diff
@@ -0,0 +1,454 @@
+		DeleteFunc: eh.deleteEvent,
+	})
+}
+
+// Handles new event objects
+func (eh *EventHandler) addEvent(obj interface{}) {
+	e, ok := obj.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", obj)
+		return
+	}
+	eventTs := getTimestamp(e)
+
+	// if event is older than the one stored in cache on startup, we've shipped it
+	if eventTs.Before(eh.InitEvent.Timestamp) {
+		return
+	}
+	// if event is equal and is in map, we've shipped it
+	if eventTs.Equal(eh.InitEvent.Timestamp) {
+		if _, ok := eh.InitEvent.RvMap[e.ResourceVersion]; ok {
+			return
+		}
+	}
+
+	labels, msg, err := eh.handleEvent(e)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Error processing event", "err", err, "event", e)
+	}
+
+	entry := newEntry(msg, eventTs, labels)
+	ok = eh.LokiClient.SendEntry(entry, eh.SendTimeout)
+	if !ok {
+		err = fmt.Errorf("error handing entry off to promtail")
+		level.Error(eh.Log).Log("err", err, "entry", entry)
+		return
+	}
+	level.Info(eh.Log).Log("msg", "Shipped entry", "eventRV", e.ResourceVersion, "eventMsg", e.Message)
+
+	// update cache with new "last" event
+	err = eh.updateLastEvent(e, eventTs)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Could not write latest event to cache", "err", err)
+	}
+}
+
+// Handles event object updates. Note that this get triggered on informer resyncs and also
+// events occurring more than once (in which case .count is incremented)
+func (eh *EventHandler) updateEvent(objOld interface{}, objNew interface{}) {
+	eOld, ok := objOld.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", eOld)
+		return
+	}
+
+	eNew, ok := objNew.(*v1.Event)
+	if !ok {
+		level.Error(eh.Log).Log("msg", "Object not v1.Event", "obj", eNew)
+		return
+	}
+
+	if eOld.GetResourceVersion() == eNew.GetResourceVersion() {
+		// ignore resync updates
+		level.Debug(eh.Log).Log("msg", "Event RV didn't change, ignoring", "eRV", eNew.ResourceVersion)
+		return
+	}
+
+	eventTs := getTimestamp(eNew)
+	// if event is older than the one stored in cache on startup, we've shipped it
+	if eventTs.Before(eh.InitEvent.Timestamp) {
+		return
+	}
+	// if event is equal and is in map, we've shipped it
+	if eventTs.Equal(eh.InitEvent.Timestamp) {
+		if _, ok := eh.InitEvent.RvMap[eNew.ResourceVersion]; ok {
+			return
+		}
+	}
+
+	labels, msg, err := eh.handleEvent(eNew)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Error processing event", "err", err, "event", eNew)
+	}
+
+	entry := newEntry(msg, eventTs, labels)
+	ok = eh.LokiClient.SendEntry(entry, eh.SendTimeout)
+	if !ok {
+		err = fmt.Errorf("error handing entry off to promtail")
+		level.Error(eh.Log).Log("err", err, "entry", entry)
+		return
+	}
+	level.Info(eh.Log).Log("msg", "Shipped entry", "eventRV", eNew.ResourceVersion, "eventMsg", eNew.Message)
+
+	err = eh.updateLastEvent(eNew, eventTs)
+	if err != nil {
+		level.Error(eh.Log).Log("msg", "Could not write latest event to cache", "err", err)
+	}
```

Comment:
This section seems a little repetitive across all the handlers, you should be able to bring it out into a single function that logs an individual event.

Reply:
Yep good call